FlowCal/plot.py

```diff
@@ -1061,6 +1302,1271 @@ def hist1d(data_list,
+        Label to use on the y axis. If None, attempts to extract channel
+        name from last data object.
+    title : str, optional
+        Plot title.
+
+    """
+
+    ###
+    # understand inputs
+    ###
+
+    # populate default input values
+    if violin_kwargs is None:
+        violin_kwargs = {'facecolor':'gray', 'edgecolor':'black'}
+
+    if draw_summary_stat_kwargs is None:
+        draw_summary_stat_kwargs = {'color':'black'}
+
+    if draw_min_line_kwargs is None:
+        draw_min_line_kwargs = {'color':'gray', 'linestyle':'--', 'zorder':-2}
+
+    if draw_max_line_kwargs is None:
+        draw_max_line_kwargs = {'color':'gray', 'linestyle':'--', 'zorder':-2}
+
+    if draw_model_kwargs is None:
+        draw_model_kwargs = {'color':'gray',
+                             'zorder':-1,
+                             'solid_capstyle':'butt'}
+
+    if draw_logx_zero_divider_kwargs is None:
+        draw_logx_zero_divider_kwargs = {'color':'gray', 'linestyle':':'}
+
+    if draw_minmax_divider_kwargs is None:
+        draw_minmax_divider_kwargs = {'color':'gray', 'linestyle':'-'}
+
+    # check x and y scales
+    if xscale not in ('linear', 'log'):
+        msg  = "`xscale` must be 'linear' or 'log'"
+        raise ValueError(msg)
+
+    if yscale not in ('linear', 'log'):
+        msg  = "`yscale` must be 'linear' or 'log'"
+        raise ValueError(msg)
+
+    # understand `data`
+    if channel is None:
+        # assume 1D sequence or sequence of 1D sequences
+        try:
+            first_element = next(iter(data))
+        except TypeError:
+            msg  = "`data` should be 1D sequence or sequence of 1D sequences."
+            msg += " Specify `channel` to use ND sequence or sequence of ND"
+            msg += " sequences."
+            raise TypeError(msg)
+
+        # promote singleton if necessary
+        try:
+            iter(first_element)  # success => sequence of 1D sequences
+            data_length = len(data)
+        except TypeError:
+            data = [data]
+            data_length = 1
+    else:
+        # assume ND sequence or sequence of ND sequences
+        try:
+            first_element               = next(iter(data))
+            first_element_first_element = next(iter(first_element))
+        except TypeError:
+            msg  = "`data` should be ND sequence or sequence of ND sequences."
+            msg += " Set `channel` to None to use 1D sequence or sequence of"
+            msg += " 1D sequences."
+            raise TypeError(msg)
+
+        # promote singleton if necessary
+        try:
+            iter(first_element_first_element)  # success => sequence of ND sequences
+            data_length = len(data)
+        except TypeError:
+            data = [data]
+            data_length = 1
+
+        # exctract channel
+        try:
+            data = [d[:,channel] for d in data]
+        except TypeError:
+            data = [[row[channel] for row in d] for d in data]
```

Comment:
What situation is this supposed to cover? If every `data[i]` is a list of lists and `channel` is an int?

Reply:
Yeah, pretty much.

In the words of the docstring, it permits `data` to be a sequence of ND sequences where the ND sequences cannot be indexed via multi-dimensional indexing (i.e. `ndseq[:,channel]`). ND sequences comprised of simpler data structures, like lists and tuples, are indeed prototypical examples.

Another data structure I had in mind was a `DataFrame`, which will *not* work as currently written. Iterating over a `DataFrame` iterates over its column names, not its rows. I figured supporting a `DataFrame` should be relegated to https://github.com/taborlab/FlowCal/issues/76, though. And I suspect a similar failure would arise with ND sequences as dictionaries mapping from channel to sequence.